Unify management of thread local settings

ilia-cher opened the pull request:
Summary:
In this PR we extend ThreadLocalState to cover dispatch keys and
ThreadLocalDebugInfo and move it from JIT interpreter down to
thread management (at::launch) and autograd (backward threads) code

Test Plan: unit tests (CI)

Differential Revision: D20615714

ngimel commented:
Should provide structure for #28520

zdevito reviewed (commented):
Changes to ThreadLocalState object look good to me. I don't have context on what ThreadLocalDebugInfo is so I will let someone else review that part.

ezyang commented:
deferring to Zachary's review. I also don't have context on ThreadLocalDebugInfo

Commit:
Debug information guard

Summary:
This diff fixes the issues with current handling of debug information passed
along the execution of the model. (For example, it is possible that multiple
calls to the debug guard may override each other)

Test Plan: CI test/cpp/jit

Differential Revision: D20602775

fbshipit-source-id: 8a7a29e8f8ffadaffb46230fd48b287147d2e43a

Commit:
Unify management of thread local settings (#35523)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/35523

In this PR we extend ThreadLocalState to cover dispatch keys and
ThreadLocalDebugInfo and move it from JIT interpreter down to
thread management (at::launch) and autograd (backward threads) code

Test Plan: unit tests (CI)

Reviewed By: dzhulgakov

Differential Revision: D20615714

fbshipit-source-id: 4911f043f11c8cfcb06ae1d7b6348fcb4ec5b03b
